src/Evolution/Systems/NewtonianEuler/BoundaryCorrections/Hll.cpp

```diff
@@ -0,0 +1,294 @@
+#include "NumericalAlgorithms/DiscontinuousGalerkin/NormalDotFlux.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+namespace NewtonianEuler::BoundaryCorrections {
+template <size_t Dim>
+Hll<Dim>::Hll(CkMigrateMessage* msg) noexcept : BoundaryCorrection<Dim>(msg) {}
+
+template <size_t Dim>
+std::unique_ptr<BoundaryCorrection<Dim>> Hll<Dim>::get_clone() const noexcept {
+  return std::make_unique<Hll>(*this);
+}
+
+template <size_t Dim>
+void Hll<Dim>::pup(PUP::er& p) {
+  BoundaryCorrection<Dim>::pup(p);
+}
+
+template <size_t Dim>
+template <size_t ThermodynamicDim>
+double Hll<Dim>::dg_package_data(
+    const gsl::not_null<Scalar<DataVector>*> packaged_mass_density,
+    const gsl::not_null<tnsr::I<DataVector, Dim, Frame::Inertial>*>
+        packaged_momentum_density,
+    const gsl::not_null<Scalar<DataVector>*> packaged_energy_density,
+    const gsl::not_null<Scalar<DataVector>*>
+        packaged_normal_dot_flux_mass_density,
+    const gsl::not_null<tnsr::I<DataVector, Dim, Frame::Inertial>*>
+        packaged_normal_dot_flux_momentum_density,
+    const gsl::not_null<Scalar<DataVector>*>
+        packaged_normal_dot_flux_energy_density,
+    const gsl::not_null<Scalar<DataVector>*>
+        packaged_largest_outgoing_char_speed,
+    const gsl::not_null<Scalar<DataVector>*>
+        packaged_largest_ingoing_char_speed,
+
+    const Scalar<DataVector>& mass_density,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& momentum_density,
+    const Scalar<DataVector>& energy_density,
+
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& flux_mass_density,
+    const tnsr::IJ<DataVector, Dim, Frame::Inertial>& flux_momentum_density,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& flux_energy_density,
+
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& velocity,
+    const Scalar<DataVector>& specific_internal_energy,
+
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_covector,
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+    /*mesh_velocity*/,
+    const std::optional<Scalar<DataVector>>& normal_dot_mesh_velocity,
+    const EquationsOfState::EquationOfState<false, ThermodynamicDim>&
+        equation_of_state) const noexcept {
+  {
+    // Compute sound speed
+    Scalar<DataVector>& sound_speed = *packaged_mass_density;
+    Scalar<DataVector>& normal_dot_velocity = *packaged_energy_density;
+    sound_speed_squared(make_not_null(&sound_speed), mass_density,
+                        specific_internal_energy, equation_of_state);
+    get(sound_speed) = sqrt(get(sound_speed));
+    dot_product(make_not_null(&normal_dot_velocity), velocity, normal_covector);
+    // Compute the largest outgoing / ingoing characteristic speeds. Note that
+    // the notion of being 'largest ingoing' means taking the most negative
+    // value given that the positive direction is outward normal.
```

Comment:
2nd sentence -> "Note that the 'largest ingoing' speed has the most negative value, because the positive direction is along the outward-directed normal"